src/verbs/queue_pair.rs

```diff
@@ -157,147 +376,177 @@ impl<'res> QueuePairBuilder<'res> {
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_DEST_QPN;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::DestinationQueuePairNumber);
         self
     }
 
     pub fn setup_rq_psn(&mut self, rq_psn: u32) -> &mut Self {
         self.attr.rq_psn = rq_psn;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_RQ_PSN;
+        self.attr_mask = self
+            .attr_mask
+            .or(QueuePairAttributeMask::ReceiveQueuePacketSequenceNumber);
         self
     }
 
     pub fn setup_sq_psn(&mut self, sq_psn: u32) -> &mut Self {
         self.attr.sq_psn = sq_psn;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_SQ_PSN;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::SendQueuePacketSequenceNumber);
         self
     }
 
     pub fn setup_max_read_atomic(&mut self, max_read_atomic: u8) -> &mut Self {
         self.attr.max_rd_atomic = max_read_atomic;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_MAX_QP_RD_ATOMIC;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::MaxReadAtomic);
         self
     }
 
     pub fn setup_max_dest_read_atomic(&mut self, max_dest_read_atomic: u8) -> &mut Self {
         self.attr.max_dest_rd_atomic = max_dest_read_atomic;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_MAX_DEST_RD_ATOMIC;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::MaxDestinationReadAtomic);
         self
     }
 
     pub fn setup_min_rnr_timer(&mut self, min_rnr_timer: u8) -> &mut Self {
         self.attr.min_rnr_timer = min_rnr_timer;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_MIN_RNR_TIMER;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::MinResponderNotReadyTimer);
         self
     }
 
     pub fn setup_timeout(&mut self, timeout: u8) -> &mut Self {
         self.attr.timeout = timeout;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_TIMEOUT;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::Timeout);
         self
     }
 
     pub fn setup_retry_cnt(&mut self, retry_cnt: u8) -> &mut Self {
         self.attr.retry_cnt = retry_cnt;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_RETRY_CNT;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::RetryCount);
         self
     }
 
     pub fn setup_rnr_retry(&mut self, rnr_retry: u8) -> &mut Self {
         self.attr.rnr_retry = rnr_retry;
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_RNR_RETRY;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::ResponderNotReadyRetryCount);
         self
     }
 
     pub fn setup_address_vector(&mut self, ah_attr: &AddressHandleAttribute) -> &mut Self {
         self.attr.ah_attr = ah_attr.attr.clone();
-        self.attr_mask |= ibv_qp_attr_mask::IBV_QP_AV;
+        self.attr_mask = self.attr_mask.or(QueuePairAttributeMask::AddressVector);
         self
     }
 }
 
 // TODO(zhp): trait for QueuePair
+
+#[inline]
+pub fn get_needed_mask(
```

Comment:
All the following functions are not meant to be invoked directly by the user. So they should not be public